[MRG] Add framereader module

kalebdfischer opened the pull request:
#### Describe the changes
Adds support for reading efficiently individual frames of a (multi-frame) image without loading the entire Pixel Data element into memory (see also https://github.com/pydicom/pydicom/pull/534, https://github.com/pydicom/pydicom/issues/1263, https://github.com/pydicom/pydicom/issues/1243).

This work is based on @hackermd 's [ImageFileReader MR](https://github.com/pydicom/pydicom/pull/1447), with the following noteworthy changes:

- renamed ImageFileReader to FrameReader and moved to its own module so that rebasing is not a pain in the future
- refactoring to allow for caching of metadata information, especially the basic offset table and to avoid parsing datasets just for FrameReader functionality
- renamed private functions _get_bot, _read_bot, _build_bot to clarify that they are for encapsulated data
- fixed basic offset table building for RLE files that lack a readable basic offset table (FrameReader can parse frames for test data files `OBXXXX1A_rle_2frame.dcm` and `rtdose_rle.dcm`)
- added some logic to accommodate files like `SC_rgb_rle.dcm` which claim to be explicit, but are actually implicit.
- wrote tests from ground up 
- added private methods for compressed/uncompressed reading to break up long modules under `if encapsulated`
- added tolerance of files that are missing NumberOfFrames data element, but have an intact basic offset table
- made log statements more informative
- removed ambiguous usage of `metadata`
- allow for non-path file-likes to be provided as input to FrameReader
- allow for specification of read_partial params `defer_size`, `force`, `specific_tags`
- added validation logic to confirm PixelData is present and dataset had attributes necessary to read frames

#### Tasks
- [x] Unit tests added that reproduce the issue or prove feature is working
- [x] Fix or feature added
- [x] Code typed and mypy shows no errors
- [x] Documentation updated (if relevant)
  - [x] No warnings during build
  - [x] Preview link (CircleCI -> Artifacts -> `doc/_build/html/index.html`)
- [x] Unit tests passing and overall coverage the same or better

kalebdfischer commented:
Not sure why mypy is complaining about files that are not changed by this MR - I haven't been able to reproduce locally.

kalebdfischer commented on pydicom/framereader.py:
```diff
@@ -0,0 +1,1132 @@
+        fragment_offsets.append(frame_position - first_frame_location)
+        first_two_bytes = fp.read(2, True)
+        if not fp.is_little_endian:
+            first_two_bytes = first_two_bytes[::-1]
```
I could remove this since it _appears_ that encapsulated should always be little endian, but leaving in case there's some other reason @hackermd included the if statement.

mrbean-bremen replied:
Yes, this is probably not needed, but maybe @hackermd can comment on this. Or generally review this PR, for that...

hackermd commented:
Thanks for all your great work @kalebdfischer! I will take a closer look in the next couple of days.

darcymason reviewed (commented):
Thanks for putting this together.  I've done a quick pass through, will try to look again in more detail later.  Just some minor comments at this point.

darcymason commented on pydicom/tests/test_framereader.py:
```diff
@@ -0,0 +1,420 @@
+    for f in get_testdata_files("*.dcm")
+    if not any([f.endswith(x) for x in cannot_read])
+]
+
```
Easier and more readable, I think, to make `can_read` a `set`, and then `all_paths=(Path(f) for f in get_testdata_files("*.dcm"))`, and `can_read_paths = [p for p in all_paths if p.name not in cannot_read]`.

Really pydicom should have a `get_testdata_paths` function since we've been trying to move more towards pathlib; something for a future PR if we can remember...

kalebdfischer replied:
Changed [here](https://github.com/pydicom/pydicom/blob/6d2c73aafea9529e756de0ab41a4d923d4ea0da3/pydicom/tests/test_framereader.py#L75)

kalebdfischer commented on pydicom/framereader.py:
```diff
@@ -0,0 +1,1132 @@
+# Copyright 2008-2022 pydicom authors. See LICENSE file for details.
+"""Utilities for parsing DICOM PixelData frames.
```
updated [here](https://github.com/pydicom/pydicom/blob/6d2c73aafea9529e756de0ab41a4d923d4ea0da3/pydicom/framereader.py#L1)

kalebdfischer commented on pydicom/framereader.py:
```diff
@@ -0,0 +1,1132 @@
+}
+_UINT_PIXEL_DATA_TAGS = {
+    0x7FE00010,
+}
```
updated [here](https://github.com/pydicom/pydicom/blob/6d2c73aafea9529e756de0ab41a4d923d4ea0da3/pydicom/framereader.py#L39)

kalebdfischer commented on pydicom/framereader.py:
```diff
@@ -0,0 +1,1132 @@
+    """
+    if fp.tell() != pixel_data_location:
+        fp.seek(pixel_data_location, 0)
+
```
Good catch - missed the discrepancy in the refactor.  Updated to validate and raise [here](https://github.com/pydicom/pydicom/blob/6d2c73aafea9529e756de0ab41a4d923d4ea0da3/pydicom/framereader.py#L100)

kalebdfischer commented on pydicom/framereader.py:
```diff
@@ -0,0 +1,1132 @@
+    ds_copy.file_meta = FileMetaDataset()
+    ds_copy.file_meta.TransferSyntaxUID = getattr(
+        original_dataset.file_meta, "TransferSyntaxUID"
+    )
```
It was lazy code that counted on the AttributeError exceptions. I've updated the [function](https://github.com/pydicom/pydicom/blob/6d2c73aafea9529e756de0ab41a4d923d4ea0da3/pydicom/framereader.py#L263) to raise ValueError for missing required attributes.

kalebdfischer commented:
> I had only a cursory glance so far, will have a closer look probably in a couple of days. Generally, I think this is a very valuable addition, and it is good to continue the work @hackermd had started here. If he did not already do this, I suggest that he also reviews this. This also needs an entry in the release notes, and ideally a description in the documentation related to reading pixel data.

I added updates to the release notes, working_with_pixel_data, and added a framereader.rst file to reference. Happy to wait for @hackermd's review.

hackermd commented on pydicom/framereader.py:
```diff
@@ -0,0 +1,1182 @@
+
+
+def decode_frame(
+        frame_bytes: bytes, original_dataset: Dataset
```
I suggest passing the individual data elements to the function rather than `original_dataset`. That is more verbose, but easier to understand.

Specifically, I suggest using the existing implementation in highdicom: https://github.com/herrmannlab/highdicom/blob/d36e2a50350fd15531f549f52dafd87587a80afe/src/highdicom/frame.py#L327-L451

If that gets implemented here in pydicom, we can just import it in highdicom.

kalebdfischer replied:
Happy to refactor with the function from highdicom, but came across an issue - this [line](https://github.com/herrmannlab/highdicom/blob/d36e2a50350fd15531f549f52dafd87587a80afe/src/highdicom/frame.py#L449) has mismatched typing. I'm not terribly comfortable just telling mypy to ignore this.

hackermd commented on pydicom/framereader.py:
```diff
@@ -0,0 +1,1182 @@
+    return ds_temp.pixel_array
+
+
+class BasicOffsetTable(List):
```
I would not make that a subclass of `List` and then add `to_dict()` and `from_dict()` methods. If you really want to implement it as a class, I suggest using a `NamedTuple` instead.

hackermd reviewed (changes requested):
I started to review this PR and added comments. However, I ultimately concluded that we should have additional design discussions.

The PR adds several classes (`FrameDataset`, `FrameInfo`, etc.) and methods (`to_dict()`) that do not directly map to DICOM information entities and are therefore problematic in my opinion. The rest of the library closely follows and directly implements the DICOM information model (e.g., `DataElement`, `Dataset`, etc.). The changes introduce via this PR divert from that approach.

Furthermore, I think the functionality mainly belongs into the `filereader` module. The main class is intended for reading frames from files (or file-like objects) and uses file-specific methods (`read()`, `seek()`, `tell()`, etc.). It is not intended to decode frames from a `Dataset` that has already been fully read into memory. In other words, it is very different from the `Dataset.pixel_array` property and similar to `dcmread()`. What differentiates it from `dcmread()` is that it provides access to individual frames and that it caches information to be reused (metadata and BOT) for repeated frame-level access. For improved cohension, I would thus add all the functionality related to file reading into the `filereader` module. Functionality related to frame decoding can remain in `frame`.

hackermd commented:
@kalebdfischer @darcymason please note that the `ImageFileReader` class is part of highdicom: https://github.com/herrmannlab/highdicom/blob/d36e2a50350fd15531f549f52dafd87587a80afe/src/highdicom/io.py#L211

I think it would be great to have that class implemented in pydicom so that we can just import it in highdicom. However, that would require that the API remains the same.

kalebdfischer commented:
> > > Thanks for all your great work @kalebdfischer! I will take a closer look in the next couple of days.
> > 
> > 
> > @hackermd thanks for the original MR! We've been maintaining a pydicom branch with the ImageFileReader for [Flywheel](https://github.com/flywheel-io)'s OHIF backend service and it's been very stable across a wide surface area of files.
> > For those using cloud storage for files, download of files is one of the greatest sources of latency. This refactor is going to enable thriftier frame retrieval the frame information can be stored and one can specifically request the frame bytes from storage on subsequent reads without need to retrieve the whole file.
> 
> Have you considered using `dicomweb_client`? It fully abstract frame-level access and supports retrieving frames via DICOMweb ([DICOMwebClient](https://dicomweb-client.readthedocs.io/en/latest/package.html#dicomweb_client.api.DICOMClient)) and reading them from local files ([DICOMfileClient](https://dicomweb-client.readthedocs.io/en/latest/package.html#dicomweb_client.api.DICOMfileClient)).

Thanks for the recommendation - our DIMSE service actually uses dicomweb_client. There are a number of reasons that pydicom is more ergonomic and practical for our OHIF backend service.

kalebdfischer commented:
> @kalebdfischer @darcymason please note that the `ImageFileReader` class is part of highdicom: https://github.com/herrmannlab/highdicom/blob/d36e2a50350fd15531f549f52dafd87587a80afe/src/highdicom/io.py#L211
> 
> I think it would be great to have that class implemented in pydicom so that we can just import it in highdicom. However, that would require the API to remain the same.

I'm happy to help with any refactoring needed by highdicom. I will take a deeper look this afternoon to ensure that this PR doesn't make the refactoring more of a pain than it needs to be.

darcymason commented:
> The rest of the library closely follows and directly implements the DICOM information model (e.g., DataElement, Dataset, etc.).

Also a good point. Pydicom has always tried to use DICOM nomenclature wherever possible.

hackermd commented on pydicom/framereader.py:
```diff
@@ -0,0 +1,1182 @@
+        return frame_dataset
+
+
+class FrameInfo:
```
Also, be aware that there are datasets with hundreds of thousands of frames and we don't want this class to become a performance bottleneck.

kalebdfischer replied:
I don't see any way that this class imposes a performance bottleneck that ImageFileReader does not.

hackermd commented:
cc'ing @pieper @fedorov @CPBridge for awareness

hackermd commented:
> > I think it would be great to have that class implemented in pydicom so that we can just import it in highdicom
> 
> It seems reasonable to me to make this as compatible as possible. If it's going to be here anyway, you might as well not have duplicate functionality in highdicom.

Yes, if the API and behavior remains the same, `highdicom.io.ImageFileReader` could just be an alias for the class in pydicom and we could deprecate the highdicom implementation.

kalebdfischer commented:
> I started to review this PR and added comments. However, I ultimately concluded that we should have additional design discussions.

Thank you for the in-depth review. I'll be working on revisions today. 

> The PR adds several classes (`FrameDataset`, `FrameInfo`, etc.) and methods (`to_dict()`) that do not directly map to DICOM information entities and are therefore problematic in my opinion. The rest of the library closely follows and directly implements the DICOM information model (e.g., `DataElement`, `Dataset`, etc.). The changes introduce via this PR divert from that approach.

Could you elaborate on what the issue is and what changes you'd like? It doesn't seem terribly reasonable to remove tooling that allows storage to and from a json-like object, but I'm happy to make improvements if there are specific suggestions.  

> Furthermore, I think the functionality mainly belongs into the `filereader` module. The main class is intended for reading frames from files (or file-like objects) and uses file-specific methods (`read()`, `seek()`, `tell()`, etc.). It is not intended to decode frames from a `Dataset` that has already been fully read into memory. In other words, it is very different from the `Dataset.pixel_array` property and similar to `dcmread()`. What differentiates it from `dcmread()` is that it provides access to individual frames and that it caches information to be reused (metadata and BOT) for repeated frame-level access. For improved cohension, I would thus add all the functionality related to file reading into the `filereader` module. Functionality related to frame decoding can remain in `frame`.

I'll defer to the maintainers on this one, but the benefits of a separate module are:
- filereader module does not get really long and cluttered
- rebasing and merging are easier to manage
- having the FrameReader class and associated functions in its own module makes the library more readable overall

hackermd commented:
> Could you elaborate on what the issue is and what changes you'd like? It doesn't seem terribly reasonable to remove tooling that allows storage to and from a json-like object, but I'm happy to make improvements if there are specific suggestions.

The library implements the DICOM standard and uses the terms defined by the [DICOM Model of the Real World](https://dicom.nema.org/medical/dicom/current/output/chtml/part03/chapter_7.html) and [DICOM Information Model](https://dicom.nema.org/medical/dicom/current/output/chtml/part03/chapter_6.html). For example, "Frame Info" is not a defined Information Entity (IE), Information Object Definition (IOD), Module, Macro, Attribute, etc.

hackermd commented:
> having the FrameReader class and associated functions in its own module makes the library more readable overall

The problem with the term `FrameReader` is that is doesn't make clear that frames are read from a file. Also the `filereader` module encapsulates all data structures and functions for reading data from files. Therefore, including the class in this module increases cohesion.

hackermd commented:
@kalebdfischer I suggest implementing the API of the [highdicom.io.ImageFileReader](https://highdicom.readthedocs.io/en/latest/package.html#highdicom.io.ImageFileReader) class. I am all for improvement of implementation details, but our goal is to avoid duplication of functionality between *pydicom* and *highdicom*. Once the class is implemented in *pydicom*, we can deprecate it in *highdicom*.

darcymason commented:
@kalebdfischer,  @hackermd, @mrbean-bremen. and perhaps @pieper @fedorov @CPBridge @scaramallion 

I'm reviewing issues again for consideration in release 2.4.  This discussion has been idle for a while, and I'm thinking this is too complex to resolve in short order, and I'd rather get the release out in the next few weeks.  So I suggest re-assigning this to the following (major) release, 3.0.  Perhaps also as part of the Roadmap development it will become clearer how pydicom should incorporate this kind of functionality. 

Any strong thoughts on this?

scaramallion commented:
Closing, implemented in `pixels` module
